src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -171,21 +244,24 @@ trait EntityComponent {
       unmarshalEntitiesWithIds(entityAttributeAction).map(_.map { case (id, entity) => entity })
     }
 
-    def unmarshalEntitiesWithIds(entityAttributeAction: ReadAction[Seq[EntityAndAttributesRawSqlQuery.EntityAndAttributesResult]]): ReadAction[Map[Long, Entity]] = {
-      entityAttributeAction.map { entityAttributeRecords =>
-        val allEntityRecords = entityAttributeRecords.map(_.entityRecord).toSet
+    def unmarshalEntitiesWithIds(entityAttributeAction: ReadAction[Seq[EntityAndAttributesRawSqlQuery.EntityAndAttributesResult]]): ReadAction[mutable.LinkedHashMap[Long, Entity]] = {
+      entityAttributeAction.map(unmarshalEntitiesWithIds)
+    }
 
-        // note that not all entities have attributes, thus the collect below
-        val entitiesWithAttributes = entityAttributeRecords.collect {
-          case EntityAndAttributesRawSqlQuery.EntityAndAttributesResult(entityRec, Some(attributeRec), refEntityRecOption) => ((entityRec.id, attributeRec), refEntityRecOption)
-        }
+    def unmarshalEntitiesWithIds(entityAttributeRecords: Seq[entityQuery.EntityAndAttributesRawSqlQuery.EntityAndAttributesResult]): mutable.LinkedHashMap[Long, Entity] = {
+      val allEntityRecords = new mutable.LinkedHashSet() ++ entityAttributeRecords.map(_.entityRecord)
 
-        val attributesByEntityId = entityAttributeQuery.unmarshalAttributes[Long](entitiesWithAttributes)
+      // note that not all entities have attributes, thus the collect below
+      val entitiesWithAttributes = entityAttributeRecords.collect {
+        case EntityAndAttributesRawSqlQuery.EntityAndAttributesResult(entityRec, Some(attributeRec), refEntityRecOption) => ((entityRec.id, attributeRec), refEntityRecOption)
+      }
 
-        allEntityRecords.map { entityRec =>
-          entityRec.id -> unmarshalEntity(entityRec, attributesByEntityId.getOrElse(entityRec.id, Map.empty))
-        }.toMap
+      val attributesByEntityId = entityAttributeQuery.unmarshalAttributes[Long](entitiesWithAttributes)
+      val result = new mutable.LinkedHashMap[Long, Entity]()
+      allEntityRecords.foreach { entityRec =>
+        result += entityRec.id -> unmarshalEntity(entityRec, attributesByEntityId.getOrElse(entityRec.id, Map.empty))
       }
+      result
```

Comment:
What about `entityAttributeRecords.zipWithIndex` and using the index as the map key instead of entity ID?